torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -0,0 +1,301 @@
+
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace data {
+template <typename Dataset, typename Sampler>
+class DataLoader {
+ public:
+  using Batch = typename Dataset::BatchType;
+  using IndexBatch = std::vector<size_t>;
+
+  /// Constructs a new `DataLoader` from a `dataset` to sample from, `options`
+  /// to configure the `DataLoader` with, and a `sampler` that specifies the
+  /// sampling strategy.
+  DataLoader(Dataset dataset, DataLoaderOptions options, Sampler sampler)
+      : options_(std::move(options)),
+        dataset_size_(dataset.size()),
+        sampler_(std::move(sampler)),
+        sequencer_(new_sequencer()) {
+    // clang-format off
+    AT_CHECK(
+        options_.batch_size <= dataset.size(),
+        "Batch size (was ", options_.batch_size, ") ",
+        "must not be larger than the dataset size (was ",
+        dataset.size(), ")");
+    // clang-format on
+
+    if (options_.workers > 0) {
+      for (size_t w = 0; w < options_.workers; ++w) {
+        // Here we copy the dataset into the worker thread closure. Each worker
+        // has its own copy of the dataset. This means the dataset must be
+        // trivially copiable, or else we don't expect more than one worker to
+        // be in use.
+        workers_.emplace_back(
+            [this, dataset] { this->worker_thread(std::move(dataset)); });
+      }
+    } else {
+      main_thread_dataset_ = torch::make_unique<Dataset>(std::move(dataset));
+    }
+  }
+
+  ~DataLoader() {
+    join();
+  }
+
+  /// Returns an iterator into the `DataLoader`. The lifetime of the iterator is
+  /// bound to the `DataLoader`. In C++ standards language, the category of the
+  /// iterator is `OutputIterator`. See
+  /// https://en.cppreference.com/w/cpp/named_req/OutputIterator for what this
+  /// means. In short: you may increment the iterator and dereference it, but
+  /// cannot go back, or step forward more than one position at a time. When the
+  /// `DataLoader` is exhausted, it will compare equal with the special
+  /// "sentinel" iterator returned by `DataLoader::end()`. Most of the time, you
+  /// should only use range-for loops to loop over the `DataLoader`, but
+  /// standard algorithms like `std::copy(dataloader.begin(), dataloader.end(),
+  /// output_iterator)`  are supported too.
+  Iterator<Batch> begin() {
+    reset();
+    return Iterator<Batch>(torch::make_unique<detail::ValidIterator<Batch>>(
+        [this] { return this->next(); }));
+  }
+
+  /// Returns a special "sentinel" iterator that compares equal with a
+  /// non-sentinel iterator once the `DataLoader` is exhausted.
+  Iterator<Batch> end() {
+    return Iterator<Batch>(
+        torch::make_unique<detail::SentinelIterator<Batch>>());
+  }
+
+  /// Joins the `DataLoader`'s worker threads and drains internal queues.
+  void join() {
+    if (joined_) {
```

Comment:
Not sure how much do we care about thread safety here? Might be at least worth documenting this.

Reply:
done